### PR TITLE
arch/risc-v: implement SBI ipi extension

### PR DESCRIPTION
**Pull request by inochisa**
Implement ipi send via SBI. This patch includes:

1. formalize NuttSBI ecall interface
2. implement NuttSBI IPI extension
3. add support for sbi_ipi_send.

Log on rv-virt:ksmp64:
[qemu-ksmp64.log](https://github.com/apache/nuttx/files/15144610/qemu-ksmp64.log)

This should fix https://github.com/apache/nuttx/pull/12178#discussion_r1581985451 and replace #12242

### CODE REVIEW

**Comment by inochisa** [comment]
@xiaoxiang781216 I think you may check this.

File: arch/risc-v/src/common/supervisor/riscv_sbi.c [hunk 1]
```diff
@@ -125,7 +106,8 @@ void riscv_sbi_set_timer(uint64_t stime_value)
 uint64_t riscv_sbi_get_time(void)
 {
 #ifdef CONFIG_NUTTSBI
-  return sbi_mcall_get_time();
+  return sbi_ecall(SBI_EXT_FIRMWARE, SBI_EXT_FIRMWARE_GET_MTIME, 0, 0,
+                   0, 0, 0, 0);
 #else
 #ifdef CONFIG_ARCH_RV64
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/common/supervisor/riscv_sbi.c:112]: let's merge `#else\n#ifdef` to `#elif defined`

File: arch/risc-v/src/common/riscv_ipi.h [hunk 2]
```diff
@@ -34,10 +34,12 @@
...
 static inline void riscv_ipi_send(int cpu)
 {
-#if defined(RISCV_IPI)
+#if defined(CONFIG_ARCH_USE_S_MODE)
+  riscv_sbi_send_ipi(0x1, cpu);
+#elif defined(RISCV_IPI)
   putreg32(1, (uintptr_t)RISCV_IPI + (4 * cpu));
 #else
-  PANIC();
+# error "No IPI support for this SoC"
```
Comment (xiaoxiang781216) [inline arch/risc-v/src/common/riscv_ipi.h:42]: two soaces after `#`

**Comment by inochisa** [comment]
> > @xiaoxiang781216 I think you may check this.
> 
> what's about the stack pointer array initialization?

I  have tried generating stack array with cmake. It works, but bl602 have a different ram layout. So I need to find a way to support it (maybe just let soc it self can generate this file too).

**Comment by xiaoxiang781216** [comment]
@inochisa but the new change doesn't have real change.

File: arch/risc-v/src/common/supervisor/riscv_sbi.c [hunk 3]
```diff
@@ -127,6 +127,12 @@ uint64_t riscv_sbi_get_time(void)
 #endif
 }
 
+uintptr_t riscv_sbi_send_ipi(uint32_t hmask, uintptr_t hbase)
+{
+  return sbi_ecall(SBI_EXT_IPI, SBI_EXT_IPI_SEND_IPI,
```
[the remark this answers is not in the document]
Comment (inochisa) [reply arch/risc-v/src/common/supervisor/riscv_sbi.c:132]: Yes, this is why I have attached a log on rv-virt:ksmp64.

**Comment by inochisa** [comment]
> @inochisa but the new change doesn't have real change.

@xiaoxiang781216 It is different. SBI IPI extension make SoCs only support S-mode (IIRC, bl808 and jh7110 are this case) work. Although RISCV_IPI is a mandatory for M-mode, it is not a good idea to require `RISCV_IPI` in S-mode as aclint is not accepted widely.

File: arch/risc-v/src/common/supervisor/riscv_sbi.c (hunk 1 above)
Comment (inochisa) [reply arch/risc-v/src/common/supervisor/riscv_sbi.c]: Thanks

File: arch/risc-v/src/common/riscv_ipi.h (hunk 2 above)
Comment (inochisa) [reply arch/risc-v/src/common/riscv_ipi.h]: OK

File: arch/risc-v/src/nuttsbi/Kconfig [hunk 4]
```diff
@@ -15,6 +15,12 @@ config NUTTSBI_HART_CNT
 	int "Amount of harts in SoC"
 	default 1
 
+config NUTTSBI_IPI_BASE
+	hex "MTIMECMP base address"
```
[the remark this answers is not in the document]
Comment (inochisa) [reply arch/risc-v/src/nuttsbi/Kconfig]: `MTIMECMP` -> `MSWI`

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**